drgn_tools/crash_net.py

```diff
@@ -0,0 +1,372 @@
+            "if_list",
+        ):
+            addr_bytes = prog.read(addr.addr.in6_u.u6_addr8.address_of_(), 16)
+            ips.append(ipaddress.IPv6Address(addr_bytes))
+    return ips
+
+
+def for_each_netdev(ns: Object) -> Iterator[Tuple[str, Object]]:
+    """
+    Iterate over each ``struct net_device *`` of a net namespace
+
+    :param ns: ``struct net *``
+    :returns: an iterator of ``struct net_device *``
+    """
+    # Since commit ff92741270bf ("net: introduce name_node struct
+    # to be used in hashlist"), net.dev_name_head is a list of name nodes. Prior
+    # to that, it contains the net_devices directly.
+    # The below approach is similar to what's already in drgn's
+    # netdev_get_by_name().
+    try:
+        entry_type = ns.prog_.type("struct netdev_name_node")
+        member = "hlist"
+        entry_is_name_node = True
+    except LookupError:
+        entry_type = ns.prog_.type("struct net_device")
+        member = "name_hlist"
+        entry_is_name_node = False
+
+    for i in range(_NETDEV_HASHENTRIES):
+        head = ns.dev_name_head[i]
+        for dev_node in hlist_for_each_entry(entry_type, head, member):
+            dev = dev_node.dev if entry_is_name_node else dev_node
+            yield (dev_node.name.string_().decode("utf-8"), dev)
+
+
+def net_crash(prog: Program) -> None:
+    """
+    Display the network devices, their names and addresses.
+    :param prog: drgn program
+    :returns: None.
+    """
+    rows = [["NET_DEVICE", "NAME", "IP ADDRESS(ES)"]]
+
+    net_namespaces = network.for_each_net(prog)
+    for net_namespace in net_namespaces:
+        for name, dev in for_each_netdev(net_namespace):
+            net_device_ptr = hex(dev.dev.platform_data)
+            ip_list = [str(a) for a in netdev_ipv4s(dev) + netdev_ipv6s(dev)]
+            ips = ", ".join(ip_list)
+            rows.append(
+                [
+                    net_device_ptr,
+                    name,
+                    ips,
+                ]
+            )
+    print_table(rows)
+
+
+def crash_a(prog: Program) -> None:
+    """
+    Displays the arp cache.
+    :param prog: drgn program
+    :returns: None.
+    """
+
+    rows = [
+        ["NEIGHBOUR", "IP ADDRESS", "HW TYPE", "HW ADDRESS", "DEVICE", "STATE"]
+    ]
+
+    nht = prog["arp_tbl"].nht
+
+    hash_shift = nht.hash_shift.value_()
+    nhash_buckets = 1 << hash_shift
+    hash_buckets = nht.hash_buckets
+
+    for i in range(nhash_buckets):
+        neigh = hash_buckets[i]
+        while neigh:
+            neigh_addr = hex(neigh.value_())[2:]
+            ip_addr = ""
+            for j in range(4):
+                ip_addr += str(neigh.primary_key[j].value_()) + "."
+            ip_addr = ip_addr[:-1]
+
+            hw = neigh.ha
+            hw_addr = ""
+            for j in range(6):
+                hw_addr += str(hex(hw[j])[2:].zfill(2)) + ":"
+            hw_addr = hw_addr[:-1]
+            dev_name = neigh.dev.name.string_().decode()
```

Comment:
I would again suggest using drgn to directly read the bytes and then passing that to the ipaddress module here. Similarly for MAC addresses, it's a bit simpler to just use string formatting rather than using hex, str, and zfill manually.
```suggestion
            neigh_addr = hex(neigh.value_())[2:]
            ip_bytes = prog.read(neigh.primary_key.address_of_(), 4)
            ip_addr = str(ipaddress.IPv4Address(ip_bytes))
            
            mac_bytes = prog.read(neigh.ha.address_of_(), 6)
            hw_addr = ":".join(f"{b:02x}" for b in mac_bytes)
            dev_name = neigh.dev.name.string_().decode()
```